crypto/rsa/rsa_lib.c

```diff
@@ -163,6 +163,139 @@ void *RSA_get_ex_data(const RSA *r, int idx)
     return CRYPTO_get_ex_data(&r->ex_data, idx);
 }
 
+/*
+ * Define a scaling constant for our fixed point arithmetic.
+ * This value must be a power of two because the base two logarithm code
+ * makes this assumption.  The exponent must also be a multiple of three so
+ * that the scale factor has an exact cube root.  Finally, the scale factor
+ * should not be so large that a multiplication of two scaled numbers
+ * overflows a 64 bit unsigned integer.
+ */
+static const unsigned int scale = 1 << 18;
+static const unsigned int cbrt_scale = 1 << (2 * 18 / 3);
+
+/* Define some constants, none exceed 32 bits */
+static const unsigned int log_2  = 0x02c5c8;    /* scale * log(2) */
+static const unsigned int log_e  = 0x05c551;    /* scale * log2(M_E) */
+static const unsigned int c1_923 = 0x07b126;    /* scale * 1.923 */
+static const unsigned int c4_690 = 0x12c28f;    /* scale * 4.690 */
+
+/*
+ * Multiply two scale integers together and rescale the result.
+ */
+static ossl_inline uint64_t mul2(uint64_t a, uint64_t b)
+{
+    /*
+     * If correct rounded is desired, enable this define.
+     * Correct rounding being:
+     *      1) if the fractional part is less than one half, round down.
+     *      2) if the fractional part is greater than one half, round up.
+     *      3) if the fractional part is exactly one half, round to even.
+     * This very slightly increases accuracy.
+     */
+#ifdef OSSL_CORRECT_ROUNDING
+    uint64_t m = a * b;
+    uint32_t f = m & (scale - 1);
+
+    m /= scale;
+    if (f == scale / 2)         /* Round to even */
+        m &= ~1;
+    else if (f >= scale / 2)    /* Round up */
+        m++;
+    return m;
+#else
+    return a * b / scale;
+#endif
+}
+
+/*
+ * Calculate the cube root of a 64 bit scaled integer.
+ * Although the cube root of a 64 bit number does fit into a 32 bit unsigned
+ * integer, this is not guaranteed after scaling, so this function has a
+ * 64 bit return.  This uses the shifting nth root algorithm.
+ */
+static uint64_t icbrt64(uint64_t x)
+{
+    uint32_t r = 0;
```

Comment:
Wouldn't this be better defined as a `uint64_t` anyway. I know a `uint32_t` should be sufficiently sized, but by making it a `uint64_t` we can avoid the two ugly casts below.

Reply:
My reasoning was that `uint32_t` will be more efficient on some platforms.  Specifically, a 32x32 bit multiplication giving a 64 bit result.  Is this efficiency a concern?

One of the casts is easy to avoid, the other not so since r<sup>2</sup> does not fit in 32 bits.

Would a floating point version of the code be acceptable?  I tried to avoid it.

Reply:
> Is this efficiency a concern?

Personally I'd prefer the clarity of avoiding the casts rather than trying to second guess how the compiler might optimise things. But it's not a big deal.

> Would a floating point version of the code be acceptable? I tried to avoid it.

I see no reason for it given what you've got here.

Reply:
A floating point version would _much_ simpler:

```
#include <math.h>

static uint16_t rsa_compute_security_bits(int n)
{
    double x = log(2), y = n * x, z = log(y);

    return (uint16_t)((1.923 * cbrt(y * z * z) - 4.69) / x + 4);
}
```